README.md
# cabextract 1.6.1

### Description:

A program to extract Microsoft Cabinet files.

(C) 2000-2015 Stuart Caie <kyzer@4u.net>  
This is free software with ABSOLUTELY NO WARRANTY.  
License: GNU GPL Version 3 - See COPYING in the source.

Cabinet (.CAB) files are a form of archive, which Microsoft use to distribute their software, and things like Windows Font Packs. The cabextract program unpacks these files.

For more information, see http://www.cabextract.org.uk/ or run the command 'cabextract --help'. This repository is a simple `fork` of that [source](https://github.com/kyz/libmspack/), with a CMake build system added, and built in WIN32/WIN64 using MSVC 2015 (msvc140).

Microsoft cabinet files should not be confused with InstallShield cabinet files. InstallShield files are generally called "_sys.cab", "data1.hdr" "data1.cab", "data2.cab" and so on, and are found in the same directory as "setup.exe". They begin with the magic characters "ISc(" rather than "MSCF". cabextract will print the message "This is probably an InstallShield file." when it finds such a file. The file "doc/magic" in the cabextract source archive includes additional file-identification rules for the UNIX file(1) command, which distinguishes between Microsoft and InstallShield cabinet files.

The project includes a `mspack` library, hence the name of the repository.

### Building:

This source now uses [CMake](https://cmake.org/) for configuration and generation of the build files of your choice.

#### In Unix:

This should be as simple as -

```
$ cd build
$ cmake .. [options]
$ make
$ [sudo] make install (if desired)
```

#### In Windows:

If for example you have [MS Visual Studio](https://www.visualstudio.com/vs/community/) (MSVC) installed -

```
> cd build
> cmake .. [options]
> cmake --build . --config Release
```

CMake also has a [GUI](https://cmake.org/runningcmake/) - Set the `source` directory, and the `binary` directory to the **build** directory. Click the [Configure], choose the generator to use. When there are no `red` items, clicking [Ok] will generate the makefiles/solution files chosen. 

Of course, if using MSVC, the generated solution file can be loaded, select the Release build, and build the project.

And of course, in Windows, MinGW, and cygwin should also be supported, but not yet tried...

### Example usage:

Extracting files from a cabinet file:  
$ cabextract wibble.cab

Extracting files from an executable which contains a cabinet file:  
$ cabextract wibble.exe  
cabextract will automatically search executables for embedded cabinets

Extracting files from a set of cabinet files; wib01.cab, wib02.cab, ...:  
$ cabextract wib01.cab  
cabextract will automatically get the names of the other files

Extracting files to a directory of your choice (in this case, 'boogie'):  
$ cabextract -d boogie wibble.cab  
cabextract will create the directory if it does not already exist

Listing files from a cabinet file:  
$ cabextract -l wibble.cab

Testing the integrity of a cabinet file, without extracting it:  
$ cabextract -t wibble.cab

Just show the file included in the CAB:  
$ cabinfo wibble.cab

Stuart Caie <kyzer@4u.net>, March 2015  
Geoff R. McLane <reports _at_ geoffair _dot_ info>, April 2017  

; eof


cabextract/ChangeLog
2017-04-06  Geoff R. McLane <reports _at_ geoffair _dot_ info>

    * Started with fork of https://www.cabextract.org.uk/ GitHub 
    repository.
    * Added a CMakeLists.txt file, to use CMake configuration and 
    makefile, solution file generation
    * Added a config.h.cmake to be processed by cmake into config.h
    * Created a 'build' directory, and added build-me.bat for convenience.
    * Port and build in WIN32/WIN64 environment.
    * Moved *.c, *.h to 'utils' directory, added conditionally to 
    a utils_lib.lib...
    * Add dirent.[c|h] for systems that do not have this.
    * Some small minor enhancement to cabinfo.c
    
2016-04-16  Stuart Caie <kyzer@cabextract.org.uk>

	* configure.ac: change my email address to kyzer@cabextract.org.uk

2015-05-10  Stuart Caie <kyzer@4u.net>

	* wince_rename, wince_info: allow the possibility that strings have
	more than one null byte. Thanks to Stefan Riegler for the suggestion.

2015-05-10  Stuart Caie <kyzer@4u.net>

	* Makefile.am: Add subdir-objects option as suggested by autoreconf.

	* configure.ac: Add AM_PROG_AR as suggested by autoreconf.

2015-02-25  Stuart Caie <kyzer@4u.net>

	* create_output_name(): Simplify code; do path checking only after
	output name has been constructed. Use towlower to lowercase UTF-8
	characters. Remove both leading '/' and '\' from cab filenames and
	change both "../" and "..\" to "xx", for Cygwin and other Cygwin-like
	environments that use both '/' and '\' as path separators. Thanks to
	Alexander Cherepanov and the MITRE team.

2015-02-18  Stuart Caie <kyzer@4u.net>

	* create_output_name(): the check against leading slashes could be
	evaded with malformed UTF-8. Thanks to Alexander Cherepanov for
	reporting the issue.

2015-01-18  Stuart Caie <kyzer@4u.net>

	* wince_rename: installed files now include their install directories.
	Thanks to Alfonso Acosta for sending a patch.

2015-01-08  Stuart Caie <kyzer@4u.net>

	* create_output_name: avoid invalid memory reads because of invalid
	UTF-8 encodings. Use the Unicode substitution character for most
	bad UTF-8 or Unicode encodings. Thanks to Hanno Böck.

	* src/cabinfo.c: stop trying to convert UTF-8 to ASCII in filename.
	Instead, prepend "UTF: " and print it to the terminal verbatim.

2011-07-19  Stuart Caie <kyzer@4u.net>

	* doc/ja/cabextract.1: Katsumi Saito updated the Japanese translation
	of cabextract's manpage to include the -t option. Thanks again!

2011-05-11  Stuart Caie <kyzer@4u.net>

	* src/cabinfo.c: changed the largefile detection header to the same
	as mspack/system.h

2011-04-27  Stuart Caie <kyzer@4u.net>

	* src/cabextract.c: changed char* to const char* in line with changes
	in libmspack. See mspack/ChangeLog for more details.

2011-04-25  Stuart Caie <kyzer@4u.net>

	* load_spanning_cabinets(): Print the error code caused by a failed
	cabinet merge, before ruining the error code by calling cabd->close().
	Thanks to Drew D'Addesio for reporting the problem.

2011-01-19  Stuart Caie <kyzer@4u.net>

	* doc/wince_cab_format.html: Identified a new file flag in the WinCE
	header file. Thanks to Paul Sokolovsky for asking me to look into it.

2010-07-19  Stuart Caie <kyzer@4u.net>

	* mspack: MS-ZIP recovery now recovers more bytes. Also, when a file
	fails to unpack because of a read error during decompression, all
	subsequent files in the same folder will give the same error, instead
	of claiming to be OK. See mspack/ChangeLog for full details.

2010-07-17  Stuart Caie <kyzer@4u.net>

	* src/cabsplit: added new program, cabsplit, which splits cabinet
	files into one cabinet file per folder. Useful for debugging.

2010-06-17  Stuart Caie  <kyzer@4u.net>

	* configure.ac: Added --external-libmspack option, which allows
	you to use the libmspack library installed on your system rather
	than the copy bundled with cabextract. Thanks to Patrice Dumas
	for the patch.

2010-06-10  Stuart Caie  <kyzer@4u.net>

	* mspack: bugs in MS-ZIP and Quantum decoders have been fixed. See
	mspack/ChangeLog for full details.

2010-05-06  Stuart Caie  <kyzer@4u.net>

	* configure.ac, cabextract.c: added support for the one-arg mkdir()
	found in the MinGW32 environment. Thanks to Kevin O'Rourke for
	reporting this.
	
2009-09-24  Stuart Caie   <kyzer@4u.net>

	*  doc/wince_cab_format.html, src/wince_info, src/wince_rename:
	The files in Windows CE cabinets are numbered after the first
	ID in the FILES section, rather than what I termed the "real ID".
	Fixed the documentation and the wince tools to match. The purpose
	of the field previously called "real ID" is now unknown. Thanks
	to Marco Schindler for discovering this.

2007-08-07  Stuart Caie   <kyzer@4u.net>

	* all: cabextract and libmspack have moved to the domain
	www.cabextract.org.uk. Updated the documentation to reflect this.
	
2007-01-20  Stuart Caie   <kyzer@4u.net>

	* create_output_name(): add UTF-8 support for 6-byte characters.
	Apparently they go that far these days.

2007-01-20  Stuart Caie   <kyzer@4u.net>

	* doc/wince_cab_format.html, src/wince_info: The REGKEYS "unknown"
	flag appears to be a boolean indicating whether variable
	substitution should take place. Thanks to Dominique Gallot for
	discovering it.

2006-06-21  Stuart Caie   <kyzer@4u.net>

	* create_output_name(): add UTF-8 support for 4-byte characters.

2006-03-01  Stuart Caie   <kyzer@4u.net>

	* process_cabinets(): on the advice of Mike Mohr, cabextract no
	longer skips entire cabinets, when create_output_filename()
	returns NULL, it only skips the affected files now.

2005-10-30  Stuart Caie   <kyzer@4u.net>

	* src/cabextract.c: added test mode to cabextract, wherein files
	aren't extracted, only passed through an MD5 checksum generator.
	cabextract then lists which files passed and which files failed.

	* process_cabinets(): fixed the problem where filters don't match
	when -d option is also used. The filters match on the full
	output-file path, including the -d directory specified. We now

libmspack/test/cabd_md5.c
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif
# if !HAVE_STRCASECMP
#  define strcasecmp strcmpi
# endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <mspack.h>
#include <sys/stat.h>
#include <dirent.h>

#include <md5_fh.h>
#include <error.h>

/**
 * Matches a cabinet's filename case-insensitively in the filesystem and
 * returns the case-correct form.
 *
 * @param origcab if this is non-NULL, the pathname part of this filename
 *                will be extracted, and the search will be conducted in
 *                that directory.
 * @param cabname the internal CAB filename to search for.
 * @return a copy of the full, case-correct filename of the given cabinet
 *         filename, or NULL if the specified filename does not exist on disk.
 */
static char *find_cabinet_file(char *origcab, char *cabname) {
    struct dirent *entry;
    struct stat st_buf;
    int found = 0, len;
    char *tail, *cab;
    DIR *dir;

    /* ensure we have a cabinet name at all */
    if (!cabname || !cabname[0]) return NULL;

    /* find if there's a directory path in the origcab */
    tail = origcab ? strrchr(origcab, '/') : NULL;
    len = (tail - origcab) + 1;

    /* allocate memory for our copy */
    if (!(cab = (char *) malloc((tail ? len : 2) + strlen(cabname) + 1))) return NULL;

    /* add the directory path from the original cabinet name, or "." */
    if (tail) memcpy(cab, origcab, (size_t) len);
    else      cab[0]='.', cab[1]='/', len=2;
    cab[len] = '\0';

    /* try accessing the cabinet with its current name (case-sensitive) */
    strcpy(&cab[len], cabname);
    if (stat(cab, &st_buf) == 0) {
	found = 1;
    }
    else {
	/* cabinet was not found, look for it in the current dir */
	cab[len] = '\0';
	if ((dir = opendir(cab))) {
	    while ((entry = readdir(dir))) {
		if (strcasecmp(cabname, entry->d_name) == 0) {
		    strcat(cab, entry->d_name);
		    found = (stat(cab, &st_buf) == 0);
		    break;
		}
	    }
	    closedir(dir);
	}
    }

    if (!found || !S_ISREG(st_buf.st_mode)) {
	/* cabinet not found, or not a regular file */
	free(cab);
	cab = NULL;
    }

    return cab;
}


int main(int argc, char *argv[]) {
    struct mscab_decompressor *cabd;
    struct mscabd_cabinet *cab, *c, *c2;
    struct mscabd_file *file;
    char *cabname, *newname;
    int err;

    setbuf(stdout, NULL);
    setbuf(stderr, NULL);

    /* if self-test reveals an error */
    MSPACK_SYS_SELFTEST(err);
    if (err) return 1;

    if (!(cabd = mspack_create_cab_decompressor(&read_files_write_md5))) {
	fprintf(stderr, "can't make decompressor\n");
	return 1;
    }

    for (argv++; (cabname = *argv); argv++) {
	printf("*** %s\n", cabname);

	if (!(cab = cabd->open(cabd, cabname))) {
	    fprintf(stderr, "cab open error: %s\n", ERROR(cabd));
	    continue;
	}

	/* prepend any spanning cabinets */
	for (c = cab; c && (c->flags & MSCAB_HDR_PREVCAB); c = c->prevcab) {
	    if (!(newname = find_cabinet_file(cabname, c->prevname))) {
		fprintf(stderr, "%s: can't find \"%s\" to prepend\n",
			cabname, c->prevname);
		break;
	    }
	    if (!(c2 = cabd->open(cabd, newname))) {
		fprintf(stderr, "%s: error opening \"%s\" for prepend: %s\n",
			cabname, newname, ERROR(cabd));
		break;
	    }
	    if (cabd->prepend(cabd, c, c2) != MSPACK_ERR_OK) {
		fprintf(stderr, "%s: error prepending \"%s\": %s\n",
			cabname, newname, ERROR(cabd));
		break;
	    }
	}

	/* append any spanning cabinets */
	for (c = cab; c && (c->flags & MSCAB_HDR_NEXTCAB); c = c->nextcab) {
	    if (!(newname = find_cabinet_file(cabname, c->nextname))) {
		fprintf(stderr, "%s: can't find \"%s\" to append\n",
			cabname, c->nextname);
		break;
	    }
	    if (!(c2 = cabd->open(cabd, newname))) {
		fprintf(stderr, "%s: error opening \"%s\" for append: %s\n",
			cabname, newname, ERROR(cabd));
		break;
	    }
	    if (cabd->append(cabd, c, c2) != MSPACK_ERR_OK) {
		fprintf(stderr, "%s: error appending \"%s\": %s\n",
			cabname, newname, ERROR(cabd));
		break;
	    }
	}

	/* extract files */
	for (file = cab->files; file; file = file->next ) {
	    if (cabd->extract(cabd, file, NULL) == MSPACK_ERR_OK) {
		printf("%s  %s\n", md5_string, file->filename);
	    }
	    else {
		fprintf(stderr, "%s: error extracting \"%s\": %s\n",
			cabname, file->filename, ERROR(cabd));
	    }
	}

	/* free all resources */
	for (c2 = cab->prevcab; c2; c2 = c2->prevcab) free((void*)c2->filename);
	for (c2 = cab->nextcab; c2; c2 = c2->nextcab) free((void*)c2->filename);
	cabd->close(cabd, cab);
    }
    mspack_destroy_cab_decompressor(cabd);
    return 0;
}


cabextract/src/cabinfo.c
/* cabinfo -- dumps useful information from cabinets
 * (C) 2000-2016 Stuart Caie <kyzer@cabextract.org.uk>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif
#include <stdio.h>
#if HAVE_SYS_TYPES_H
# include <sys/types.h>
#endif
#if HAVE_STDLIB_H
# include <stdlib.h>
#endif
#if HAVE_STRING_H
# include <string.h>
#endif
#if HAVE_STRINGS_H
# include <strings.h>
#endif

#ifdef HAVE_FSEEKO
# define FSEEK fseeko
# define FTELL ftello
# define FILELEN off_t
#else
# define FSEEK fseek
# define FTELL ftell
# define FILELEN long
#endif

#if ((defined(_FILE_OFFSET_BITS) && _FILE_OFFSET_BITS >= 64) || \
     (defined(FILESIZEBITS)      && FILESIZEBITS      >= 64) || \
     (defined(SIZEOF_OFF_T)      && SIZEOF_OFF_T      >= 8)  || \
     defined(_LARGEFILE_SOURCE) || defined(_LARGEFILE64_SOURCE))
# define FL "lld"
#else
# define FL "ld"
#endif


/* structure offsets */
#define cfhead_Signature         (0x00)
#define cfhead_CabinetSize       (0x08)
#define cfhead_FileOffset        (0x10)
#define cfhead_MinorVersion      (0x18)
#define cfhead_MajorVersion      (0x19)
#define cfhead_NumFolders        (0x1A)
#define cfhead_NumFiles          (0x1C)
#define cfhead_Flags             (0x1E)
#define cfhead_SetID             (0x20)
#define cfhead_CabinetIndex      (0x22)
#define cfhead_SIZEOF            (0x24)
#define cfheadext_HeaderReserved (0x00)
#define cfheadext_FolderReserved (0x02)
#define cfheadext_DataReserved   (0x03)
#define cfheadext_SIZEOF         (0x04)
#define cffold_DataOffset        (0x00)
#define cffold_NumBlocks         (0x04)
#define cffold_CompType          (0x06)
#define cffold_SIZEOF            (0x08)
#define cffile_UncompressedSize  (0x00)
#define cffile_FolderOffset      (0x04)
#define cffile_FolderIndex       (0x08)
#define cffile_Date              (0x0A)
#define cffile_Time              (0x0C)
#define cffile_Attribs           (0x0E)
#define cffile_SIZEOF            (0x10)
#define cfdata_CheckSum          (0x00)
#define cfdata_CompressedSize    (0x04)
#define cfdata_UncompressedSize  (0x06)
#define cfdata_SIZEOF            (0x08)

/* flags */
#define cffoldCOMPTYPE_MASK            (0x000f)
#define cffoldCOMPTYPE_NONE            (0x0000)
#define cffoldCOMPTYPE_MSZIP           (0x0001)
#define cffoldCOMPTYPE_QUANTUM         (0x0002)
#define cffoldCOMPTYPE_LZX             (0x0003)
#define cfheadPREV_CABINET             (0x0001)
#define cfheadNEXT_CABINET             (0x0002)
#define cfheadRESERVE_PRESENT          (0x0004)
#define cffileCONTINUED_FROM_PREV      (0xFFFD)
#define cffileCONTINUED_TO_NEXT        (0xFFFE)
#define cffileCONTINUED_PREV_AND_NEXT  (0xFFFF)
#define cffile_A_RDONLY                (0x01)
#define cffile_A_HIDDEN                (0x02)
#define cffile_A_SYSTEM                (0x04)
#define cffile_A_ARCH                  (0x20)
#define cffile_A_EXEC                  (0x40)
#define cffile_A_NAME_IS_UTF           (0x80)


FILE *fh;
char *filename = 0;
FILELEN filelen;
void search();
void getinfo();

static int verbose = 0;

#define EndGetI32(a)  ((((a)[3])<<24)|(((a)[2])<<16)|(((a)[1])<<8)|((a)[0]))
#define EndGetI16(a)  ((((a)[1])<<8)|((a)[0]))
#define GETLONG(n) EndGetI32(&buf[n])
#define GETWORD(n) EndGetI16(&buf[n])
#define GETBYTE(n) ((int)buf[n])

#define GETOFFSET      (FTELL(fh))
#define READ(buf,len)  if (myread((void *)(buf),(len))) return
#define SKIP(offset)   if (myseek((offset),SEEK_CUR)) return
#define SEEK(offset)   if (myseek((offset),SEEK_SET)) return


int myread(void *buf, int length) {
  FILELEN remain = filelen - GETOFFSET;
  if (length > remain) length = (int) remain;
  if (fread(buf, 1, length, fh) != length) {
    perror(filename);
    return 1;
  }
  return 0;
}

int myseek(FILELEN offset, int mode) {
  if (FSEEK(fh, offset, mode) != 0) {
    perror(filename);
    return 1;
  }
  return 0;
}

void showVersion()
{
    printf("Cabinet information dumper version %s\n", VERSION);
}

const char *justExe(const char *rt)
{
    const char *exe = rt;
    size_t ii, len = strlen(rt);
    int c;
    for (ii = 0; ii < len; ii++) {
        c = rt[ii];
        if ((c == '\\') || (c == '/')) {
            if (rt[ii + 1])
                exe = &rt[ii + 1];
        }
    }
    return exe;
}


void showUsage(const char *exe)
{
    showVersion();
    printf("\n");
    printf("Usage:\n");
    printf("%s [options] <file.cab>\n", justExe(exe));
    printf("\n");
    printf("Options:\n");
    printf(" --help     (-h or -?) = This help and exit(0)\n");
    printf(" --version        (-v) = Show version and exit(0)\n");
    printf(" --VERBOSE        (-V) = Show more details, and data sections.\n");
    printf("\n");
    printf("Authors:\n");
    printf("Original 1.4 version by Stuart Caie <kyzer@4u.net>, circa 2011-05-11\n");
    printf("Windows port, and CMake build by Geoff R. McLane <reports _at_ geoffair _dot_ info>\n");
    printf("See 'AUTHORS' file, in the source, for further information\n");
    printf("\n");
    printf("License:\n");
    printf("GNU GPL Version 3. See 'COPYING' file, in source for text.\n");
    printf("\n");
}

int parseArgs(int argc, char *argv[])
{
    int i, c;
    char *arg, *sarg;
    for (i = 1; i < argc; i++)
    {
        arg = argv[i];
        c = *arg;
        if ((c == '-') || (c == '/')) {
            sarg = &arg[1];
            while (*sarg == '-')
                sarg++;
            c = *sarg;
            switch (c)
            {
            case 'h':
            case '?':
                showUsage(argv[0]);
                return 2;
            case 'v':
                showVersion();
                return 2;
            case 'V':
                verbose++;
                break;
            default:
                printf("Unknown argument '%s'! Try -?\n", arg);
                return 1;
            }
        }
        else {
            if (filename) {
                printf("Already have file name '%s'! What is this '%s'?\n", filename, arg);
                return 1;
            }
            filename = strdup(arg);
        }
    }
    if (!filename) {
        showUsage(argv[0]);
        printf("Error: No file name found in command!\n");
        return 1;
    }
    return 0;
}

int main(int argc, char *argv[]) 
{
    int iret = 0;
  if (argc <= 1) {
        showUsage(argv[0]);
    return 1;
  }
    iret = parseArgs(argc, argv);
    if (iret) {
        if (iret == 2)
            iret = 0;
        return iret;
    }

    fh = fopen(filename, "rb");
    if (!fh) {
        printf("Error: Failed to open '%s'\n", filename);
        // perror(filename);
    return 1;
  }

  if (FSEEK(fh, 0, SEEK_END) != 0) {
        printf("Error: Seek to file '%s' end FAILED!\n", filename);
        //perror(filename);
    fclose(fh);
    return 1;
  }

  filelen = FTELL(fh);
  if (FSEEK(fh, 0, SEEK_SET) != 0) {
        printf("Error: Seek to file '%s' begin FAILED!\n", filename);
        //perror(filename);
    fclose(fh);
    return 1;
  }

  printf("Examining file \"%s\" (%" FL " bytes)...\n", filename, filelen);
  search();
  fclose(fh);
  return 0;
}


#define SEARCH_SIZE (32*1024)
unsigned char search_buf[SEARCH_SIZE];

void search() {
  unsigned char *pstart = &search_buf[0], *pend, *p;
  FILELEN offset, caboff, cablen, foffset, length;
  unsigned long cablen32, foffset32;
  int state = 0;

  for (offset = 0; offset < filelen; offset += length) {
    /* search length is either the full length of the search buffer,
     * or the amount of data remaining to the end of the file,
     * whichever is less.
     */
    length = filelen - offset;
    if (length > SEARCH_SIZE) length = SEARCH_SIZE;

    /* fill the search buffer with data from disk */
    SEEK(offset);
    READ(&search_buf[0], length);
    /* read through the entire buffer. */
    p = pstart;
    pend = &search_buf[length];
    while (p < pend) {
      switch (state) {
	/* starting state */
      case 0:
	/* we spend most of our time in this while loop, looking for
	 * a leading 'M' of the 'MSCF' signature
	 */
	while (*p++ != 0x4D && p < pend);
	if (p < pend) state = 1; /* if we found tht 'M', advance state */
	break;
	
	/* verify that the next 3 bytes are 'S', 'C' and 'F' */
      case 1: state = (*p++ == 0x53) ? 2 : 0; break;
      case 2: state = (*p++ == 0x43) ? 3 : 0; break;
      case 3: state = (*p++ == 0x46) ? 4 : 0; break;
	
	/* we don't care about bytes 4-7 */
	/* bytes 8-11 are the overall length of the cabinet */
      case 8:  cablen32  = *p++;       state++; break;
      case 9:  cablen32 |= *p++ << 8;  state++; break;
      case 10: cablen32 |= *p++ << 16; state++; break;
      case 11: cablen32 |= *p++ << 24; state++; break;
	
	/* we don't care about bytes 12-15 */
	/* bytes 16-19 are the offset within the cabinet of the filedata */
      case 16: foffset32  = *p++;       state++; break;
      case 17: foffset32 |= *p++ << 8;  state++; break;
      case 18: foffset32 |= *p++ << 16; state++; break;
      case 19: foffset32 |= *p++ << 24;
	/* now we have recieved 20 bytes of potential cab header. */
	/* work out the offset in the file of this potential cabinet */
	caboff = offset + (p-pstart) - 20;
	/* check that the files offset is less than the alleged length
	 * of the cabinet, and that the offset + the alleged length are
	 * 'roughly' within the end of overall file length
	 */
	foffset = (FILELEN) foffset32;
	cablen  = (FILELEN) cablen32;
	if ((foffset < cablen) &&
	    ((caboff + foffset) < (filelen + 32)) &&
	    ((caboff + cablen) < (filelen + 32)) )
	{
	  /* found a potential result - try loading it */
	  printf("Found cabinet header at offset %" FL "\n", caboff);
	  SEEK(caboff);
	  getinfo();
	  offset = caboff + cablen;
	  length = 0;
	  p = pend;
	}
	state = 0;
	break;

      default:
	p++, state++; break;
      } /* switch state */
    } /* while p < pend */
  } /* while offset < filelen */
}

#define CAB_NAMEMAX (1024)

void getinfo() {
  unsigned char buf[64];
  unsigned char namebuf[CAB_NAMEMAX];
  char *name;

  FILELEN offset, base_offset, files_offset, base;
  int num_folders, num_files, num_blocks = 0;
  int header_res = 0, folder_res = 0, data_res = 0;
  int i, x, y;

  base_offset = GETOFFSET;

  READ(&buf, cfhead_SIZEOF);

  x = GETWORD(cfhead_Flags);

  printf(
    "\n*** HEADER SECTION ***\n\n"
    "Cabinet signature      = '%4.4s'\n"
    "Cabinet size           = %u bytes\n"
    "Offset of files        = %" FL "\n"
    "Cabinet format version = %d.%d\n"
    "Number of folders      = %u\n"
    "Number of files        = %u\n"
    "Header flags           = 0x%04x%s%s%s\n"
    "Set ID                 = %u\n"
    "Cabinet set index      = %u\n",

    buf,
    GETLONG(cfhead_CabinetSize),
    files_offset = (GETLONG(cfhead_FileOffset) + base_offset),
    GETBYTE(cfhead_MajorVersion),
    GETBYTE(cfhead_MinorVersion),
    num_folders = GETWORD(cfhead_NumFolders),
    num_files = GETWORD(cfhead_NumFiles),
    x,
    ((x & cfheadPREV_CABINET)    ? " PREV_CABINET"    : ""),
    ((x & cfheadNEXT_CABINET)    ? " NEXT_CABINET"    : ""),
    ((x & cfheadRESERVE_PRESENT) ? " RESERVE_PRESENT" : ""),
    GETWORD(cfhead_SetID),
    GETWORD(cfhead_CabinetIndex)
  );

  if (num_folders == 0) { printf("ERROR: no folders\n"); return; }
  if (num_files == 0) { printf("ERROR: no files\n"); return; }

  if (buf[0]!='M' || buf[1]!='S' || buf[2]!='C' || buf[3]!='F')
    printf("WARNING: cabinet doesn't start with MSCF signature\n");

  if (GETBYTE(cfhead_MajorVersion) > 1
  || GETBYTE(cfhead_MinorVersion) > 3)
    printf("WARNING: format version > 1.3\n");



  if (x & cfheadRESERVE_PRESENT) {
    READ(&buf, cfheadext_SIZEOF);
    header_res = GETWORD(cfheadext_HeaderReserved);
    folder_res = GETBYTE(cfheadext_FolderReserved);
    data_res   = GETBYTE(cfheadext_DataReserved);
  }

  printf("Reserved header space  = %u\n", header_res);
  printf("Reserved folder space  = %u\n", folder_res);
  printf("Reserved datablk space = %u\n", data_res);

  if (header_res > 60000)
    printf("WARNING: header reserved space > 60000\n");

  if (header_res) {
    printf("[Reserved header: offset %" FL ", size %u]\n", GETOFFSET,
	   header_res);
    SKIP(header_res);
  }

  if (x & cfheadPREV_CABINET) {
    base = GETOFFSET;
    READ(&namebuf, CAB_NAMEMAX);
    SEEK(base + strlen((char *) namebuf) + 1);
    printf("Previous cabinet file  = %s\n", namebuf);
    if (strlen((char *) namebuf) > 256) printf("WARNING: name length > 256\n");

    base = GETOFFSET;
    READ(&namebuf, CAB_NAMEMAX);
    SEEK(base + strlen((char *) namebuf) + 1);
    printf("Previous disk name     = %s\n", namebuf);
    if (strlen((char *) namebuf) > 256) printf("WARNING: name length > 256\n");
  }

  if (x & cfheadNEXT_CABINET) {
    base = GETOFFSET;
    READ(&namebuf, CAB_NAMEMAX);
    SEEK(base + strlen((char *) namebuf) + 1);
    printf("Next cabinet file      = %s\n", namebuf);
    if (strlen((char *) namebuf) > 256) printf("WARNING: name length > 256\n");

    base = GETOFFSET;
    READ(&namebuf, CAB_NAMEMAX);
    SEEK(base + strlen((char *) namebuf) + 1);
    printf("Next disk name         = %s\n", namebuf);
    if (strlen((char *) namebuf) > 256) printf("WARNING: name length > 256\n");
  }

  printf("\n*** FOLDERS SECTION ***\n");

  for (i = 0; i < num_folders; i++) {
    offset = GETOFFSET;
    READ(&buf, cffold_SIZEOF);

    switch(GETWORD(cffold_CompType) & cffoldCOMPTYPE_MASK) {
    case cffoldCOMPTYPE_NONE:    name = "stored";  break;
    case cffoldCOMPTYPE_MSZIP:   name = "MSZIP";   break;
    case cffoldCOMPTYPE_QUANTUM: name = "Quantum"; break;
    case cffoldCOMPTYPE_LZX:     name = "LZX";     break;
    default:                     name = "unknown"; break;
    }

    printf(
      "\n[New folder at offset %" FL "]\n"
      "Offset of folder       = %" FL "\n"
      "Num. blocks in folder  = %u\n"
      "Compression type       = 0x%04x [%s]\n",

      offset,
      base_offset + GETLONG(cffold_DataOffset),
      GETWORD(cffold_NumBlocks),
      GETWORD(cffold_CompType),
      name
    );

    num_blocks += GETWORD(cffold_NumBlocks);

    if (folder_res) {
      printf("[Reserved folder: offset %" FL ", size %u]\n", GETOFFSET,
	     folder_res);
      SKIP(folder_res);
    }
  }

  printf("\n*** FILES SECTION *** Count %d\n", num_files);

  if (verbose) {
      printf("Showing verbose details of each section...\n");
  } else {
      printf("Use -V to show more than 'date time size name' for each.\n");
  }

  if (GETOFFSET != files_offset) {
    printf("WARNING: weird file offset in header\n");
    SEEK(files_offset);
  }


  for (i = 0; i < num_files; i++) {
    offset = GETOFFSET;
    READ(&buf, cffile_SIZEOF);

    switch (GETWORD(cffile_FolderIndex)) {
    case cffileCONTINUED_PREV_AND_NEXT:
      name = "continued from previous and to next cabinet";
      break;
    case cffileCONTINUED_FROM_PREV:
      name = "continued from previous cabinet";
      break;
    case cffileCONTINUED_TO_NEXT:
      name = "continued to next cabinet";
      break;
    default:
      name = "normal folder";
      break;
    }
    
    x = GETWORD(cffile_Attribs);
    
    base = GETOFFSET;
    READ(&namebuf, CAB_NAMEMAX);
    SEEK(base + strlen((char *) namebuf) + 1);
    if (strlen((char *) namebuf) > 256) printf("WARNING: name length > 256\n");
    if (verbose) {
    printf(
      "\n[New file at offset %" FL "]\n"
      "File name              = %s%s\n"
      "File size              = %u bytes\n"
      "Offset within folder   = %u\n"
      "Folder index           = 0x%04x [%s]\n"
      "Date / time            = %02d/%02d/%4d %02d:%02d:%02d\n"
      "File attributes        = 0x%02x %s%s%s%s%s%s\n",
      offset,
      x & cffile_A_NAME_IS_UTF ? "UTF: " : "",
      namebuf,
      GETLONG(cffile_UncompressedSize),
      GETLONG(cffile_FolderOffset),
      GETWORD(cffile_FolderIndex),
      name,
       GETWORD(cffile_Date)       & 0x1f,
      (GETWORD(cffile_Date) >> 5) & 0xf,
      (GETWORD(cffile_Date) >> 9) + 1980,
       GETWORD(cffile_Time) >> 11,
      (GETWORD(cffile_Time) >> 5) & 0x3f,
      (GETWORD(cffile_Time) << 1) & 0x3e,
      x,
      (x & cffile_A_RDONLY)      ? "RDONLY " : "",
      (x & cffile_A_HIDDEN)      ? "HIDDEN " : "",
      (x & cffile_A_SYSTEM)      ? "SYSTEM " : "",
      (x & cffile_A_ARCH)        ? "ARCH "   : "",
      (x & cffile_A_EXEC)        ? "EXEC "   : "",
      (x & cffile_A_NAME_IS_UTF) ? "UTF-8"   : ""
    );
  }
    else
    {
        printf("%02d/%02d/%4d %02d:%02d %12u  %s\n",
            GETWORD(cffile_Date) & 0x1f,
            (GETWORD(cffile_Date) >> 5) & 0xf,
            (GETWORD(cffile_Date) >> 9) + 1980,
            GETWORD(cffile_Time) >> 11,
            (GETWORD(cffile_Time) >> 5) & 0x3f,
            GETLONG(cffile_UncompressedSize),
            namebuf
        );

    }
  }

  printf("\n*** DATABLOCKS SECTION *** Count %d\n", num_blocks);

  if (verbose) {

  printf("*** Note: offset is BLOCK offset. Add 8 for DATA offset! ***\n\n");

  for (i = 0; i < num_blocks; i++) {
    offset = GETOFFSET;
    READ(&buf, cfdata_SIZEOF);
    x = GETWORD(cfdata_CompressedSize);
    y = GETWORD(cfdata_UncompressedSize);
    printf("Block %6d: offset %12" FL " / csum %08x / c=%5u / u=%5u%s\n",
	   i, offset, GETLONG(cfdata_CheckSum), x, y,
	   ((x > (32768+6144)) || (y > 32768)) ? " INVALID" : "");
    SKIP(x);
  }
  }
  else {
      printf("Use -V to show more details of the data sections\n");
  }
}

// eof
